lib/Dialect/Triton/Transforms/RewriteTMAToLegacy.cpp

```diff
@@ -0,0 +1,428 @@
+ * using casted tensor descriptors and offsets and casting the results back to
+ * tensor pointers.
+ *
+ * We have special handling for TMA loads/stores and the make tensor descriptor
+ * op.
+ *
+ * @note Why use the conversion pattern rewriter? In most cases the defining
+ * operation of a tensor descriptor will be a make tensor descriptor op.
+ * However, this isn't always true - for example, if the tensor descriptor is a
+ * function argument or is in a conditional statement, we need better tracking
+ * of the pointer, shape, and strides.
+ */
+class RewriteTritonTmaToLegacyPass
+    : public impl::TritonRewriteTMAToLegacyBase<RewriteTritonTmaToLegacyPass> {
+  void runOnOperation() override {
+    auto op = getOperation();
+
+    mlir::ConversionTarget target(getContext());
+    target.addDynamicallyLegalDialect<mlir::arith::ArithDialect,
+                                      mlir::scf::SCFDialect,
+                                      mlir::triton::TritonDialect>(
+        [](mlir::Operation *op) {
+          return !hasATensorDescriptorType(op->getOperandTypes()) &&
+                 !hasATensorDescriptorType(op->getResultTypes());
+        });
+    target.addDynamicallyLegalOp<triton::FuncOp>([](triton::FuncOp func_op) {
+      return !hasATensorDescriptorType(func_op.getFunctionType().getInputs()) &&
+             !hasATensorDescriptorType(func_op.getFunctionType().getResults());
+    });
+    target.addLegalOp<mlir::UnrealizedConversionCastOp>();
+
+    mlir::TypeConverter converter;
+
+    converter.addConversion([](mlir::Type t) {
+      // Most types don't require any conversion
+      return t;
+    });
+    converter.addConversion([](mlir::triton::TensorDescType t,
+                               llvm::SmallVectorImpl<mlir::Type> &out) {
+      // We convert a tensor descriptor into an pointer, and a shape and stride
+      // for each dimension, i.e., we create 1+2*rank values. Note that tensor
+      // descriptors may be signed/unsigned integers whereas pointers should
+      // always be signless.
+      auto tensor_type = t.getBlockType();
+      out.push_back(triton::getPointerType(
+          toSignlessIntegerType(tensor_type.getElementType())));
+      out.insert(out.end(), 2 * tensor_type.getRank(),
+                 mlir::IntegerType::get(t.getContext(), 64));
+      return mlir::success();
+    });
+
+    mlir::RewritePatternSet patterns(op->getContext());
+
+    // Populate conversion patterns to handle loops, function calls, and arith
+    // ops.
+    triton::populateFunctionTypeConversions(converter, patterns);
+    mlir::scf::populateSCFStructuralTypeConversions(converter, patterns);
+    mlir::arith::populateArithTypeConversions(converter, patterns);
+
+    patterns
+        .add<RewriteMakeTensorDesc, RewriteLoadPattern, RewriteStorePattern>(
+            converter, &getContext());
+
+    ConversionConfig config;
+    config.buildMaterializations = false;
+
+    if (mlir::failed(mlir::applyPartialConversion(
```

Comment:
Don't we need a full conversion here, otherwise it will fail at a later point in the compilation?
```suggestion
    if (mlir::failed(mlir::applyFullConversion(
```

Reply:
I think `applyPartialConversion()` is correct. We only want to change operations that have tensor descriptors operands/results and we have explicitly marked these as illegal so `applyPartialConversion()` will still fail if there are any tensor descriptors left in the graph. We also want to leave all other existing operations - it looks like `applyFullConversion()` would require us to mark all other operations as legal if we wanted to keep them.

Reply:
Shouldn't everything already be legal because you've used dynamically legal dialect above?

Reply:
> we have explicitly marked these as illegal so applyPartialConversion() will still fail if there are any tensor descriptors left in the graph.

Actually, yeah you're totally right. So this seems good.